src/backend/distributed/utils/statistics_collection.c

```diff
@@ -138,6 +147,173 @@ CollectBasicUsageStatistics(void)
 }
 
 
+/* CheckForUpdates queries Citus servers for newer releases of Citus. */
+void
+CheckForUpdates(void)
+{
+	StringInfo url = makeStringInfo();
+	appendStringInfoString(url, STATS_COLLECTION_HOST "/v1/releases/latest?edition=");
+
+	if (!UrlEncode(url, CITUS_EDITION))
+	{
+		ereport(WARNING, (errmsg("url encoding '%s' failed", CITUS_EDITION)));
+		return;
+	}
+
+	if (!SendHttpGetJsonRequest(url->data, HTTP_TIMEOUT_SECONDS,
+								&CheckForUpdatesCallback))
+	{
+		ereport(WARNING, (errmsg("checking for updates failed")));
+	}
+}
+
+
+/*
+ * CheckForUpdatesCallback receives the response for the request sent by
+ * CheckForUpdates(). It processes the response, and if there is a newer release
+ * of Citus available, logs a LOG message. This function returns 0 if there are
+ * any errors in the received response, which means we didn't consume the data.
+ * Otherwise, it returns (size * count) which means we consumed all of the data.
+ */
+static size_t
+CheckForUpdatesCallback(char *contents, size_t size, size_t count, void *userData)
+{
+	const int32 citusVersionMajor = CITUS_VERSION_NUM / 10000;
+	const int32 citusVersionMinor = (CITUS_VERSION_NUM / 100) % 100;
+	const int32 citusVersionPatch = CITUS_VERSION_NUM % 100;
+	Jsonb *responseJsonb = NULL;
+	StringInfo releaseVersion = makeStringInfo();
+	int32 releaseMajor = 0;
+	int32 releaseMinor = 0;
+	int32 releasePatch = 0;
+	char *updateType = NULL;
+	MemoryContext savedContext = CurrentMemoryContext;
+
+	StringInfo responseNullTerminated = makeStringInfo();
+	appendBinaryStringInfo(responseNullTerminated, contents, size * count);
+
+	/* jsonb_in can throw errors */
+	PG_TRY();
+	{
+		responseJsonb = DatumGetJsonb(
```

Comment:
Again, I think this is too many operations for a line; if the line breaks like this, better to just break it up.

Reply:
Done.